### PR TITLE
instanceof

### PR DESCRIPTION
**Pull request by lukyanoffpashok**

### CODE REVIEW

File: aibolit/patterns/instanceof/instance_of.py [hunk 1]
```diff
@@ -15,16 +15,32 @@ def __traverse_node(self, node):
...
-        for _, node_elem in node.filter(javalang.tree.MethodInvocation):
-            if node_elem.member == 'isInstance':
-                lines.append(node_elem.position.line)
-
+        depth = 0
+        cur_line = 0
+        bin_depth = -10
+        m_depth = -10
+        for _, destination, edge_type in dfs_labeled_edges(node.tree, node.root):
+            cur_node = node.tree.nodes[destination]
```
Comment (aravij) [inline aibolit/patterns/instanceof/instance_of.py:23]: Seems like a redundant line.

File: aibolit/patterns/instanceof/instance_of.py [hunk 2]
```diff
@@ -15,16 +15,32 @@ def __traverse_node(self, node):
...
-            if node_elem.operator == 'instanceof' and node_elem.operandl.position is not None:
-                code_line = node_elem.operandl.position.line or node_elem.operandr.position.line
-                lines.append(code_line)
-        for _, node_elem in node.filter(javalang.tree.MethodInvocation):
-            if node_elem.member == 'isInstance':
-                lines.append(node_elem.position.line)
-
+        depth = 0
+        cur_line = 0
+        bin_depth = -10
```
Comment (aravij) [inline aibolit/patterns/instanceof/instance_of.py:20]: What this magical numbers means?

File: aibolit/patterns/instanceof/instance_of.py [hunk 3]
```diff
@@ -15,16 +15,32 @@ def __traverse_node(self, node):
...
+        m_depth = -10
+        for _, destination, edge_type in dfs_labeled_edges(node.tree, node.root):
+            cur_node = node.tree.nodes[destination]
+            if edge_type == 'forward':
+                cur_node = node.tree.nodes[destination]
+                node_type = cur_node['type']
+                if 'source_code_line' in cur_node:
+                    cur_line = cur_node['source_code_line']
+                if node_type == ASTNodeType.BINARY_OPERATION:
```
Comment (aravij) [inline aibolit/patterns/instanceof/instance_of.py:29]: Functionality to get string representation of binary operation could be moved to general AST class to be used by everyone.

File: aibolit/patterns/instanceof/instance_of.py [hunk 4]
```diff
@@ -15,16 +15,32 @@ def __traverse_node(self, node):
...
+                    bin_depth = depth
+                if node_type == ASTNodeType.STRING and cur_node['string'] == 'instanceof' and depth - bin_depth == 1:
+                    lines.append(cur_line)
+                    bin_depth = -10
+                if node_type == ASTNodeType.METHOD_INVOCATION:
```
Comment (aravij) [inline aibolit/patterns/instanceof/instance_of.py:34]: Checking for binary operations and method invocations could be split into two different function.

File: aibolit/patterns/instanceof/instance_of.py [hunk 5]
```diff
@@ -15,16 +15,32 @@ def __traverse_node(self, node):
...
         lines = []
-        for _, node_elem in node.filter(javalang.tree.BinaryOperation):
-            if node_elem.operator == 'instanceof' and node_elem.operandl.position is not None:
-                code_line = node_elem.operandl.position.line or node_elem.operandr.position.line
-                lines.append(code_line)
-        for _, node_elem in node.filter(javalang.tree.MethodInvocation):
-            if node_elem.member == 'isInstance':
-                lines.append(node_elem.position.line)
-
+        depth = 0
```
Comment (aravij) [inline aibolit/patterns/instanceof/instance_of.py:18]: You could move `_traverse_node` to `value` function.

**Comment by 0crat** [comment]
This pull request [#424](https://github.com/yegor256/aibolit/issues/424) is assigned to @aravij[/z](https://www.0crat.com/u/aravij), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/865c5888-87eb-43ab-b734-0ac2566c9d37); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: aibolit/patterns/instanceof/instance_of.py (hunk 1 above)
Comment (lukyanoffpashok) [reply aibolit/patterns/instanceof/instance_of.py]: removed it

File: aibolit/patterns/instanceof/instance_of.py (hunk 2 above)
Comment (lukyanoffpashok) [reply aibolit/patterns/instanceof/instance_of.py]: Initial depth for binary operation. You can use any number <= 0. Changed -10 to 0.

File: aibolit/patterns/instanceof/instance_of.py (hunk 5 above)
Comment (lukyanoffpashok) [reply aibolit/patterns/instanceof/instance_of.py:19]: ok

File: aibolit/patterns/instanceof/instance_of.py (hunk 3 above)
Comment (lukyanoffpashok) [reply aibolit/patterns/instanceof/instance_of.py:29]: I don't use full string representation of binary operation

File: aibolit/patterns/instanceof/instance_of.py (hunk 4 above)
Comment (lukyanoffpashok) [reply aibolit/patterns/instanceof/instance_of.py:34]: Are you sure that it makes sense?

**Comment by aravij** [comment]
@rultor  merge

**Comment by aravij** [comment]
@acheshkov please merge)

**Comment by acheshkov** [comment]
@rultor merge

**Pull request merged**